File: demos/julia/fft_demo.jl
# demos/julia/fft_demo.jl
# Demonstrates: FFT signal processing on CPU and GPU.

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "build", "bindings", "julia"))

using Insight

function separator(title)
    println("\n" * "="^40)
    println("  $title")
    println("="^40)
end

function run_fft_cpu()
    separator("CPU FFT")

    # Generate a signal: sum of two sinusoids
    n = 64
    t_vals = collect(0:n-1) ./ n
    signal = sin.(2π * 5 .* t_vals) .+ 0.5 .* sin.(2π * 12 .* t_vals)

    x = Insight.from_data(Float32.(signal))
    first8 = [string(round(Insight.item(x, i-1), digits=3)) for i in 1:8]
    println("Input signal (first 8): " * join(first8, " ") * " ...")

    # RFFT -> IRFFT roundtrip
    X = Insight.fft(x)
    println("FFT output length: $(Insight.numel(X)) ()")

    x_recon = Insight.ifft(X)
    recon8 = [string(round(Insight.item(x_recon, i-1), digits=3)) for i in 1:8]
    println("Reconstructed signal (first 8): " * join(recon8, " ") * " ...")

    # Check reconstruction error (ifft returns complex, extract real part)
    x_data = Insight.to_data(x)
    x_recon_real = Insight.real_part(x_recon)
    x_recon_data = Insight.to_data(x_recon_real)
    max_err = maximum(Base.abs.(x_data .- x_recon_data))
    println("Max reconstruction error: $max_err")

    # F64 FFT
    separator("CPU FFT (F64)")
    signal_f64 = cos.(2π * 3 .* collect(0:n-1) ./ n)
    x64 = Insight.from_data(signal_f64, Insight.float64)
    X64 = Insight.fft(x64)
    x64_recon = Insight.ifft(X64)
    x64_data = Insight.to_data(x64)
    x64_recon_real = Insight.real_part(x64_recon)
    x64_recon_data = Insight.to_data(x64_recon_real)
    max_err64 = maximum(Base.abs.(x64_data .- x64_recon_data))
    println("F64 FFT roundtrip max error: $max_err64")

    # next_fast_len
    separator("FFT Length Optimization")
    println("next_fast_len(64)  = $(Insight.next_fast_len(64))")
    println("next_fast_len(100) = $(Insight.next_fast_len(100))")
    println("next_fast_len(127) = $(Insight.next_fast_len(127))")
end

function run_fft_gpu()
    n = 64
    t_vals = collect(0:n-1) ./ n
    signal = sin.(2π * 5 .* t_vals) .+ 0.5 .* sin.(2π * 12 .* t_vals)

    x_cpu = Insight.from_data(Float32.(signal))
    x = Insight.to(x_cpu, 1)  # GPUPlace
    X = Insight.fft(x)
    x_recon = Insight.to(Insight.ifft(X), 0)  # CPUPlace

    recon8 = [string(round(Insight.item(x_recon, i-1), digits=3)) for i in 1:8]
    println("GPU FFT->IFFT roundtrip (first 8): " * join(recon8, " ") * " ...")

    x_recon_data = Insight.to_data(Insight.real_part(x_recon))
    signal_data = Float64.(signal)
    max_err = maximum(Base.abs.(signal_data .- Float64.(x_recon_data)))
    println("GPU max reconstruction error: $max_err")

    # GPU F64 FFT
    separator("GPU FFT (F64)")
    signal_f64 = cos.(2π * 3 .* collect(0:n-1) ./ n)
    x64_cpu = Insight.from_data(signal_f64, Insight.float64)
    x64 = Insight.to(x64_cpu, 1)
    X64 = Insight.fft(x64)
    x64_recon = Insight.to(Insight.ifft(X64), 0)
    x64_recon_data = Insight.to_data(Insight.real_part(x64_recon))
    max_err64 = maximum(Base.abs.(signal_f64 .- x64_recon_data))
    println("GPU F64 FFT roundtrip max error: $max_err64")
end

println("Insight7 FFT Demo (Julia)")

run_fft_cpu()

if Insight.has_device(1)
    separator("GPU FFT")
    try
        run_fft_gpu()
    catch
    end
end

println("\nDone!")

File: demos/python/fft_demo.py
"""FFT demo for Insight7 Python bindings.

Demonstrates: FFT signal processing (RFFT/IRFFT roundtrip) on CPU
and GPU (if available), with F32 and F64 precision.
"""

import sys
import os
import math

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "bindings", "python"))
sys.path.insert(
    0, os.path.join(os.path.dirname(__file__), "..", "..", "build", "bindings", "python")
)

import insight as ins
import numpy as np


def separator(title):
    print(f"\n{'=' * 40}")
    print(f"  {title}")
    print(f"{'=' * 40}")


def gpu_available():
    try:
        ins.load_backend("cuda")
        return True
    except Exception:
        return False


def run_fft_cpu():
    separator("CPU FFT")

    # Generate a signal: sum of two sinusoids
    n = 64
    t = np.arange(n) / n
    signal = np.sin(2 * math.pi * 5 * t) + 0.5 * np.sin(2 * math.pi * 12 * t)
    signal = signal.astype(np.float32)

    x = ins.from_numpy(signal)
    print(f"Input signal (first 8): {x[:8]}")

    # RFFT -> IRFFT roundtrip
    X = ins.rfft(x)
    print(f"RFFT output length: {X.numel()} (complex)")

    x_recon = ins.irfft(X, n=n)
    print(f"Reconstructed signal (first 8): {x_recon[:8]}")

    # Check reconstruction error
    max_err = np.max(np.abs(signal - x_recon.numpy()))
    print(f"Max reconstruction error: {max_err:.6e}")

    # F64 FFT
    separator("CPU FFT (F64)")
    signal_f64 = np.cos(2 * math.pi * 3 * np.arange(n) / n).astype(np.float64)
    x64 = ins.from_numpy(signal_f64)
    X64 = ins.rfft(x64)
    x64_recon = ins.irfft(X64, n=n)
    max_err_f64 = np.max(np.abs(signal_f64 - x64_recon.numpy()))
    print(f"F64 FFT roundtrip max error: {max_err_f64:.6e}")

    # next_fast_len
    separator("FFT Length Optimization")
    print(f"next_fast_len(64)  = {ins.next_fast_len(64)}")
    print(f"next_fast_len(100) = {ins.next_fast_len(100)}")
    print(f"next_fast_len(127) = {ins.next_fast_len(127)}")


def run_fft_gpu():
    separator("GPU FFT")

    n = 64
    t = np.arange(n) / n
    signal = (np.sin(2 * math.pi * 5 * t) + 0.5 * np.sin(2 * math.pi * 12 * t)).astype(np.float32)

    x = ins.from_numpy(signal).to(ins.GPUPlace(0))
    X = ins.rfft(x)
    x_recon = ins.irfft(X, n=n).to(ins.CPUPlace())

    print(f"GPU RFFT->IRFFT roundtrip (first 8): {x_recon[:8]}")
    max_err = np.max(np.abs(signal - x_recon.numpy()))
    print(f"GPU max reconstruction error: {max_err:.6e}")

    # GPU F64 FFT
    separator("GPU FFT (F64)")
    signal_f64 = np.cos(2 * math.pi * 3 * np.arange(n) / n).astype(np.float64)
    x64 = ins.from_numpy(signal_f64).to(ins.GPUPlace(0))
    X64 = ins.rfft(x64)
    x64_recon = ins.irfft(X64, n=n).to(ins.CPUPlace())
    max_err_f64 = np.max(np.abs(signal_f64 - x64_recon.numpy()))
    print(f"GPU F64 FFT roundtrip max error: {max_err_f64:.6e}")


def main():
    ins.init()

    print("Insight7 FFT Demo (Python)")

    run_fft_cpu()

    if gpu_available():
        try:
            run_fft_gpu()
        except Exception:
            pass

    print("\nDone!")


if __name__ == "__main__":
    main()
